### PR TITLE
Updating User API Doc (part 1)

### PR DESCRIPTION
**Pull request by madpink**
The issue: https://github.com/HabitRPG/habitica/issues/8087

I've updated the API documentation for the first batch if User

Includes: GET /api/v3/user – POST /api/v3/user/buy/:key

----
UUID: b0413351-405f-416f-8787-947ec1c85199

### CODE REVIEW

File: website/server/controllers/api-v3/user.js
Comment (TheHollidayInn) [inline website/server/controllers/api-v3/user.js]: http://apidocjs.com/#param-api-success-example

Comment (madpink) [reply website/server/controllers/api-v3/user.js]: The example was so long that I thought the description would be better. Would an annotated example be acceptable?

It'll still be a hundred lines or so...

```
{
  "success": true,
  "data": {
    "_id": "11be077e-4a54-4d3a-96b6-8f14c101aeff",
    "migration": "20170131_habit_birthday.js",
    "loginIncentives": 9,
    "webhooks": [],
    "pushDevices": [],
    "extra": {},
    "tasksOrder": {
      "rewards": [--TASKS LISTED BY ID],
      "todos": [--TASKS LISTED BY ID],
      "dailys": [--TASKS LISTED BY ID],
      "habits": [--TASKS LISTED BY ID],
    },
    "inbox": {
      "optOut": false,
      "messages": {},
      "blocks": [],
      "newMessages": 0
    },
    "tags": [
      {
        "name": "Work",
        "id": "94d3a62b-2f69-4b75-b930-d93dd2c21038"
      }
    ],
    "notifications": [    ],
    "stats": {
        --stats, training, buffs--    },
    "profile": {
      "name": "Ezra"
    },
;

Comment (madpink) [reply website/server/controllers/api-v3/user.js]: on the other hand, I'm looking at what MathWiz did with Content, I could just empty out most of the sub-keys like below... is that too little information?

```{
  "success": true,
  "data": {
    "_id": "11be077e-4a54-4d3a-96b6-8f14c101aeff",
    "migration": "20170131_habit_birthday.js",
    "loginIncentives": 9,
    "webhooks": [],
    "pushDevices": [],
    "extra": {},
    "tasksOrder": {
      "rewards": [ ],
      "todos": [ ],
      "dailys": [],
      "habits": [ ]
    },
    "inbox": {},
    "tags": [],
    "notifications": [    ],
    "stats": {
        --stats, training, buffs--    },
    "profile": {
      "name": "Sisyphus"
    },
    "preferences": {},
    "party": {},
    "guilds": [],
    "invitations": {},
    "challenges": [],
    "newMessages": {},
    "lastCron": "2017-02-06T12:24:56.080Z",
    "items": {},
    "history": {},
    "flags": {},
    "purchased": { },
    "balance": 0,
    "contributor": {},
    "backer": {},
    "achievements": {},
    "_v": 286,
    "auth": { },
    "id": "11be077e-4a54-4d3a-96b6-8f14c101aeff"},
  "notifications": []
}

Comment (Alys) [reply website/server/controllers/api-v3/user.js]: My personal feeling is that it would be enough to display something like this:

```
{
  "success": true,
  "data": {
    "_id": "11be077e-4a54-4d3a-96b6-8f14c101aeff",
    [All applicable user data goes here. Fetch data from your own account as an example. See website/server/models/user/* for details.]
  }
}
```

A partial list could be misleading, and the full list is too large as you noted, and is updated too often as paglias mentioned. Even a partial list is likely to become outdated fairly quickly and we'd almost always forget to update it. Anyone who wants to know what's in the data can test it  easily enough with their own account.

However if other admins disagree, I'm fine with whatever they suggest. :)

Comment (madpink) [reply website/server/controllers/api-v3/user.js]: so if the success example is abbreviated, should I remove the apiDescription?

Comment (Alys) [reply website/server/controllers/api-v3/user.js]: No, I really like your apiDescription. It's a useful overview that describes the things that aren't obvious, with no significant technical details that might become outdated, and the "including (but not limited to)" part is great for making it clear that could be more items.

Comment (TheHollidayInn) [reply website/server/controllers/api-v3/user.js]: I'm for linking to the User model as a reference

Comment (madpink) [inline website/server/controllers/api-v3/user.js]: do you mean http://habitica.wikia.com/wiki/User ?

Comment (madpink) [reply website/server/controllers/api-v3/user.js]: do you mean this?
 https://github.com/HabitRPG/habitica/tree/develop/website/server/models/user

Comment (Alys) [reply website/server/controllers/api-v3/user.js]: Ah heck, I wrote this days ago, but left it as pending instead of sending it. :(

@madpink Yes https://github.com/HabitRPG/habitica/tree/develop/website/server/models/user . If we end up using something like the message that I suggested above for the `data` object, that full link should replace the directory path that I used - same content but the full link is definitely better. Otherwise, it would be a good addition to the apiDescription.

Comment (TheHollidayInn) [reply website/server/controllers/api-v3/user.js]: Yep! Sorry for the late reply :P

Comment (madpink) [reply website/server/controllers/api-v3/user.js]: ☕️  Updated

**Comment by madpink** [comment]
Changed the API description to a link to the user model directory, also cleaned up trailing spaces (oops)

File: website/server/controllers/api-v3/user.js
Comment (Alys) [reply website/server/controllers/api-v3/user.js]: @madpink, really sorry that I didn't send that review earlier. :(  This was part of it, in response to your question "so if the success example is abbreviated, should I remove the apiDescription?"

"No, I really like your apiDescription. It's a useful overview that describes the things that aren't obvious, with no significant technical details that might become outdated, and the "including (but not limited to)" part is great for making it clear that could be more items."

I still think the details you put in under (and including)  "@apiDescription The user profile contains data related to the authenticated user including (but not limited to)" were very useful. Linking to the model as well is great, but the brief, plain-English summary of the top level items would be much-loved by coders new to the model. It's exactly the sort of high-level overview that makes it easier to start understanding the details of the code. If you'd like to put back all the points you wrote, I think it would be great! But if you think it's too late and you'd rather not, I'll understand. :)

Comment (madpink) [reply website/server/controllers/api-v3/user.js]: It's never too late...  I put the description back in, and I moved the link to the model into an example

**Comment by madpink** [comment]
I have another batch of User doc updates, should I add them to this PR or wait for this to be accepted and make a new one?

**Comment by Alys** [comment]
@madpink Create a new issue. Thanks for asking! That's what @crookedneighbor was getting at when he said to do at most 10 routes at a time. Splitting them up makes reviewing them less overwhelming. :)

**Comment by madpink** [comment]
I am not sure how to submit a separate PR for the same file... can anyone advise?

(Or do I just wait for this to be merged?)

**Comment by Alys** [comment]
@madpink Ah, might be best to wait for the previous one to be merged to prevent conflicts. Sorry about that.

Although to answer your question, you'd submit the PR the same way as the first one, but you'd probably have to fix conflicts before the second one could be merged.

However your question makes me wonder if you're using the same branch for all you PRs. Another sign that you might be doing that is that you're using develop for this one. :) While that will work for one PR at a time, as soon as you want to work on two changes at once, it causes problems. It's recommended to create a new branch for every piece of work you do. For example:
```
   git fetch upstream
   git checkout -b 2017-02-17-api-docs-user-2 upstream/develop
   .... do work here then commit or stash ...
   git fetch upstream
   git checkout -b 2017-02-21-some-other-change upstream/develop
```
The two checkout lines create two different branches, each based on upstream/develop (using git fetch to make sure your local copy of upstream/develop is up to date). You can then work in each branch without affecting the other one. The naming scheme for your branches is entirely up to you; I like using dates so I can easily find ones that I've forgotten to delete when they're finished.

**Comment by madpink** [comment]
thanks @Alys ... I think I will stick with waiting :)

File: website/server/controllers/api-v3/user.js
Comment (crookedneighbor) [inline website/server/controllers/api-v3/user.js]: You don't need to include the notifications array in these examples.

**Review by crookedneighbor — CHANGES_REQUESTED** [review]

File: website/server/controllers/api-v3/user.js
Comment (crookedneighbor) [inline website/server/controllers/api-v3/user.js]: Is this right? We send a boolean?

File: website/server/controllers/api-v3/user.js
Comment (crookedneighbor) [inline website/server/controllers/api-v3/user.js:617]: Just checked, it is a boolean.

**Review by crookedneighbor — APPROVED** [review]

**Comment by crookedneighbor** [comment]
As soon as the build passes, I'm fine with merging this.

**Comment by madpink** [comment]
is there anything else that I can help with at this time?

**Comment by TheHollidayInn** [comment]
Looks like you have lint issues

**Review by madpink — COMMENTED** [review]
Removed a trailing space that was causing problem

**Comment by paglias** [comment]
Notated towards your next tier

**Comment by madpink** [comment]
I have the next batch of changes ready, at what point should I add them?

**Comment by paglias** [comment]
@madpink you can open a PR whenever you prefer

**Pull request merged**
